### PR TITLE
Feature/show actor in history

### PR DESCRIPTION
**Pull request by edumoreira1506**
Added actor of the change in history of the task. Preview:
Peek 2019-12-06 10-06

### CODE REVIEW

File: app/assets/javascripts/components/stories/History/HistoryChanges.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+          <div>
+            <i className="mi md-18">arrow_forward</i>
+          </div>
+          <div> {value[1] || "---"} </div>
+        </div>
+      </div>
+    );
+  });
+
+  return <div> {result} </div>;
```
Comment (nolleto) [inline app/assets/javascripts/components/stories/History/HistoryChanges.js]: I would create a child component `HistoryChange` and render a list of `HistoryChange`.
Also, you could put `value[0]` and `value[1]` in a variable with a meaningful name.

Comment (nolleto) [reply app/assets/javascripts/components/stories/History/HistoryChanges.js]:
```js
const HistoryChange = ({ title, value0, value1 }) => (
  <div className="info">
    <div className="action">
      {title}
    </div>
    <div className="changes">
      <div>{value0}</div>
      <div>
        <i className="mi md-18">arrow_forward</i>
      </div>
      <div>{value1} </div>
    </div>
  </div>
);

HistoryChange.propTypes = {
  value0: PropTypes.string?, 
  value1: PropTypes.string?,
}

HistoryChange.propTypes = {
  value0: '---', 
  value1: '---',
}

const getTitle = (propName) => I18n.t(`activerecord.attributes.story.${propName}`)

const HistoryChanges = ({ changes }) => 
  mapObject(changes, (val, key) => (
    <HistoryChange title={getTitle(key)} value0={val[0]} value1={val[1]}/>
  ))
```
https://underscorejs.org/#mapObject

File: app/assets/javascripts/components/stories/History/index.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
+import React from "react";
+import PropTypes from "prop-types";
+import HistoryHeader from './HistoryHeader';
+import HistoryChanges from './HistoryChanges';
+
+const History = ({ history }) => {
+  const array = history.sort(
```
Comment (nolleto) [inline app/assets/javascripts/components/stories/History/index.js]: You may use [`sortBy`](https://underscorejs.org/#sortBy) here 🤔 
```js
const array = sortBy(history, ({ activity }) => activity.created_at);
```
Check if you will get the same result.
Also, you can extract it to a function:
```js
const sortByActivityCreation = histories => sortBy(histories, ({ activity }) => activity.created_at);
```

File: app/assets/javascripts/components/stories/History/index.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    (a, b) => new Date(b.activity.created_at) - new Date(a.activity.created_at)
+  );
+  const result = array.map(item => {
+    return (
+      <div key={item.activity.id} className="history-activity" data-id="history-activity">
```
Comment (nolleto) [inline app/assets/javascripts/components/stories/History/index.js]: You can create a component for it.
`HistoryItem` maybe 🤔

File: app/assets/javascripts/models/beta/history.js [hunk 4]
```diff
@@ -0,0 +1,9 @@
+export const getDate = (key, changes) => {
+  if (key.endsWith("_at")) {
+    return [
+      changes[key][0] ? I18n.l("date.formats.long", changes[key][0]) : null,
```
Comment (nolleto) [inline app/assets/javascripts/models/beta/history.js]: You can create a variable with `changes[key][0]` and `changes[key][1]` with meaningful names.
I believe `changes` is an `Object` where each value is an `Array` with the previous value and the next value so:

```js
export const getDate = (key, changes) => {
  const [previousValue, nextValue] = changes[key];
  
  if (key.endsWith("_at")) {  
    return [
      previousValue ? I18n.l("date.formats.long", previousValue) : null,
      previousValue ? I18n.l("date.formats.long", nextValue) : null
    ];
  }

  return [previousValue, nextValue];
};
```

File: app/assets/javascripts/components/stories/History/HistoryChanges/HistoryChange.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+HistoryChange.defaultProps = {
+  oldValue: '---',
+  newValue: '---'
+};
+
+HistoryChange.propTypes = {
+  oldValue: PropTypes.oneOfType([
+    PropTypes.number,
+    PropTypes.string
```
Comment (talyssonoc) [inline app/assets/javascripts/components/stories/History/HistoryChanges/HistoryChange.js]: Why can this be either a number or a string?

File: app/assets/javascripts/components/stories/History/index.js [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+import PropTypes from "prop-types";
+import HistoryItem from './HistoryItem';
+import { sortByActivityCreation } from '../../../models/beta/history';
+
+const History = ({
+  history
+}) =>
+  <div className="History">
+    {
+      sortByActivityCreation(history).reverse().map(item =>
```
Comment (talyssonoc) [inline app/assets/javascripts/components/stories/History/index.js]: This shouldn't be inside the component. It's not performant to sort and reverse it in every re-render and it's not the component's responsibility to do it.

My suggestion: do it when the data is received (inside the method called by the action that fetches it) from the backend and save it already sorted and reversed in the store.

File: app/assets/javascripts/components/stories/History/HistoryChanges/index.js [hunk 7]
```diff
@@ -0,0 +1,20 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import HistoryChange from './HistoryChange';
+import * as History from '../../../../models/beta/history';
+
+const HistoryChanges = ({ changes }) =>
+  History.unserializeChanges(changes).map(change =>
```
Comment (talyssonoc) [inline app/assets/javascripts/components/stories/History/HistoryChanges/index.js]: This should be done inside the method called by the action as well, it's not a component's responsibility.

Also, side note: it'd be better to use `deserialize` instead of `unserialize` :smile:

File: app/assets/javascripts/models/beta/history.js [hunk 8]
```diff
@@ -0,0 +1,22 @@
+import { sortBy } from 'underscore';
+
+export const getDate = (key, changes) => {
+  const [oldValue, newValue] = changes[key];
```
Comment (talyssonoc) [inline app/assets/javascripts/models/beta/history.js]: Instead of passing changes and key, what do you think  about passing the old and new value directly? It'll make this function less coupled with the `changes` object.

File: app/assets/stylesheets/new_board/_history.scss [hunk 9]
```diff
@@ -1,38 +1,74 @@
...
-
-    .changes {
+  
+    .header {
       display: flex;
-      font-style: italic;
-
+      justify-content: space-between;
+  
+      .title {
```
Comment (talyssonoc) [inline app/assets/stylesheets/new_board/_history.scss]: Can we use BEM selectors here? I know you weren't the one who created those but we can refactor it now.

File: app/assets/stylesheets/new_board/_key_frames.scss [hunk 10]
```diff
@@ -0,0 +1,8 @@
+@keyframes heightAnimation {
```
Comment (talyssonoc) [inline app/assets/stylesheets/new_board/_key_frames.scss]: This name sounds kinda generic, can we make it more clear about what it does exactly? Also, the `98%` value reminded me of this post https://dev.to/maxart2501/percentages-in-css-you-re-using-them-wrong-kpn, maybe this can  be improved as well using a different unity?

File: spec/controllers/stories/activities_controller_spec.rb [hunk 11]
```diff
@@ -30,7 +30,7 @@
       end
 
       it 'returns all the activities from a Story' do
-        expect(response.body).to eq(assigns[:activities].to_json)
+        expect(response.body).to eq(assigns[:activities].to_json(include: :user))
```
Comment (talyssonoc) [inline spec/controllers/stories/activities_controller_spec.rb]: Why was this needed?

File: spec/javascripts/components/stories/history/history_changes/history_change_spec.js [hunk 12]
```diff
@@ -0,0 +1,44 @@
+import React from "react";
+import { shallow } from "enzyme";
+import HistoryChange from "../../../../../../app/assets/javascripts/components/stories/History/HistoryChanges/HistoryChange";
```
Comment (talyssonoc) [inline spec/javascripts/components/stories/history/history_changes/history_change_spec.js]: We probably can do it like this here:

```suggestion
import HistoryChange from "components/stories/History/HistoryChanges/HistoryChange";
```

File: spec/javascripts/components/stories/history/history_item_spec.js [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+  it('renders the component', () => {
+    const { wrapper } = render();
+
+    expect(wrapper).toExist();
+  });
+
+  it('renders header', () => {
+    const { header } = render();
+
+    expect(header).toExist();
```
Comment (talyssonoc) [inline spec/javascripts/components/stories/history/history_item_spec.js]: Should some specific text be present in the `header`? Or in `changes`? We should assert it as well :smile:

File: spec/javascripts/models/beta/history_spec.js [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+    });
+
+    describe('when key is state', () => {
+      const key = "state";
+      const changes = {
+        state: ["finished", "delivered"]
+      };
+
+      it("return just the changes", () => {
+        expect(History.getDate(key, changes)[0]).toEqual("finished");
```
Comment (talyssonoc) [inline spec/javascripts/models/beta/history_spec.js]: What do you think about doing something like that:

```suggestion
        expect(History.getDate(key, changes)).toEqual(["finished", "delivered"]);
```

File: app/controllers/stories/activities_controller.rb [hunk 15]
```diff
@@ -4,6 +4,6 @@ def index
     @story = @project.stories.find(params[:story_id])
     @activities = policy_scope(Activity).by_story(@story.id)
 
-    render json: @activities
+    render json: @activities.as_json(include: :user)
```
Comment (talyssonoc) [inline app/controllers/stories/activities_controller.rb]: Is this needed? V1 board shows the author without this change, maybe we can simplify it somehow so we don't need to send in the payload some data that we already have in  the frontend.

File: app/assets/javascripts/components/stories/History/HistoryChanges/HistoryChange.js (hunk 5 above)
Comment (edumoreira1506) [reply app/assets/javascripts/components/stories/History/HistoryChanges/HistoryChange.js]: In the most part of the cases is a `string`, like change the description, story type, etc, but when the estimate is changed, the `oldValue` will be a number. Example:
Screenshot from 2019-12-10 10-02-00

Estimate is number, story type and updated at is a string.

File: app/assets/javascripts/models/beta/history.js (hunk 8 above)
Comment (edumoreira1506) [reply app/assets/javascripts/models/beta/history.js]: Sure! I liked of idea

File: app/assets/stylesheets/new_board/_history.scss (hunk 9 above)
Comment (edumoreira1506) [reply app/assets/stylesheets/new_board/_history.scss]: Sure!

File: spec/javascripts/components/stories/history/history_item_spec.js (hunk 13 above)
Comment (edumoreira1506) [reply spec/javascripts/components/stories/history/history_item_spec.js]: Actually the `header` and `changes` are other components that have her tests too:

https://github.com/Codeminer42/cm42-central/pull/637/files#diff-3c49478dd272270b344271d346564073R1

https://github.com/Codeminer42/cm42-central/pull/637/files#diff-3c49478dd272270b344271d346564073R1

And there, this are tested.

File: app/assets/javascripts/actions/story.js [hunk 16]
```diff
@@ -108,13 +108,14 @@ export const confirmBeforeSaveIfNeeded = async (story, confirm, needConfirmation
...
-  async (dispatch, getState, { Story }) => {
-    const { stories } = getState();
+  async (dispatch, getState, { Story, History }) => {
+    const { stories, users } = getState();
     const story = Story.findById(Story.withScope(stories, from), storyId)
     dispatch(loadHistory(story.title))
     try {
       const activities = await Story.getHistory(story.id, story.projectId)
-      dispatch(receiveHistory(activities, story.title))
+      const unserializedActivities = History.deserializeHistory(activities, users)
```
Comment (talyssonoc) [inline app/assets/javascripts/actions/story.js]: What do you think about moving this call one step inner and put it inside `Story.getHistory`? It's ok to pass `users` to this method.

File: app/assets/javascripts/components/stories/History/HistoryChanges/HistoryChange.js (hunk 5 above)
Comment (talyssonoc) [reply app/assets/javascripts/components/stories/History/HistoryChanges/HistoryChange.js]: Can we create a shape for that then?

File: app/assets/javascripts/models/beta/history.js [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+  });
+
+export const sortByActivityCreation = history =>
+  sortBy(history, ({ activity }) => activity.created_at);
+
+export const deserializeHistory = (history, users) =>
+  sortByActivityCreation(history).map(item => ({
+    activity: {
+      ...item.activity,
+      changes: [...deserializeChanges(item.activity.subject_changes)],
```
Comment (talyssonoc) [inline app/assets/javascripts/models/beta/history.js]: We probably don't need to spread inside an array here. Can we make `changes: deserializeChanges(/* arguments */)`?

File: app/assets/javascripts/actions/story.js (hunk 16 above)
Comment (edumoreira1506) [reply app/assets/javascripts/actions/story.js]: For sure.

File: app/assets/javascripts/actions/story.js [hunk 18]
```diff
@@ -108,12 +108,12 @@ export const confirmBeforeSaveIfNeeded = async (story, confirm, needConfirmation
 }
 
 export const showHistory = (storyId, from) =>
-  async (dispatch, getState, { Story }) => {
-    const { stories } = getState();
+  async (dispatch, getState, { Story, History }) => {
```
Comment (talyssonoc) [inline app/assets/javascripts/actions/story.js]:
```suggestion
  async (dispatch, getState, { Story }) => {
```

File: app/assets/javascripts/components/stories/History/HistoryChanges/HistoryChange.js [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+      <div data-id="history-old-value"> { oldValue } </div>
+      <div>
+        <i className="mi md-18">arrow_forward</i>
+      </div>
+      <div data-id="history-new-value"> { newValue } </div>
+    </div>
+  </div>
+
+HistoryChange.defaultProps = {
+  oldValue: '---',
```
Comment (talyssonoc) [inline app/assets/javascripts/components/stories/History/HistoryChanges/HistoryChange.js]: I think these fallbacks are not working. They will only be considered if the value won't actually be passed, but if you call this component with `undefined`/`null` it'll use `undefined`/`null` instead of falling back to the `defaultProps'.

Also, can we hide the changes in ids? Like the "Owned by id", we can hide that since the "Owned by" is already shown. We also don't need to show the change for the name initials since the complete name is already there :smile: 

Screenshot from 2019-12-11 09-10-22

File: app/assets/javascripts/models/beta/history.js [hunk 20]
```diff
@@ -0,0 +1,28 @@
...
+
+export const sortByActivityCreation = history =>
+  sortBy(history, ({ activity }) => activity.created_at);
+
+export const deserializeHistory = (history, users) =>
+  sortByActivityCreation(history).map(item => ({
+    activity: {
+      ...item.activity,
+      changes: deserializeChanges(item.activity.subject_changes),
+      user: users.find(user => user.id === item.activity.user_id).name
```
Comment (talyssonoc) [inline app/assets/javascripts/models/beta/history.js]: What do you think  about creating a model `User` and have a method `User.findById(users, id)`?

File: app/assets/javascripts/store/index.js [hunk 21]
```diff
@@ -25,7 +26,8 @@ const dependencies = {
   PastIteration,
   Notification,
   UrlService,
-  Search
+  Search,
+  History
```
Comment (talyssonoc) [inline app/assets/javascripts/store/index.js]: I think we won't need that here anymore now?

File: app/assets/javascripts/models/beta/history.js (hunk 20 above)
Comment (edumoreira1506) [reply app/assets/javascripts/models/beta/history.js]: Sure!

**Comment by talyssonoc** [comment]
Found this missing translation when I removed the requester of a story

Screenshot from 2019-12-11 13-54-26

**Comment by talyssonoc** [comment]
Can we change it to be just "Requested by" instead of "Requested by name" like we to with "Owned by"?

Screenshot from 2019-12-11 16-12-47

File: app/assets/javascripts/models/beta/history.js [hunk 22]
```diff
@@ -0,0 +1,38 @@
...
+  ];
+
+const isDate = key => key.endsWith("_at");
+
+const isId = key => key.endsWith("_id");
+
+const isInitials = key => key === 'owned_by_initials';
+
+export const deserializeChanges = changes =>
+  Object.keys(changes).map(key => {
```
Comment (nolleto) [inline app/assets/javascripts/models/beta/history.js]: I believe you can you [mapObject](https://underscorejs.org/#mapObject) here.

File: app/assets/javascripts/models/beta/history.js [hunk 23]
```diff
@@ -0,0 +1,38 @@
...
+      activity: {
+        ...item.activity,
+        changes: filterChanges(deserializeChanges(item.activity.subject_changes)),
+        user: findById(users, item.activity.user_id).name
+      }
+    }))
+    .reverse()
+
+export const filterChanges = changes =>
+  changes.filter(item => !isInitials(item.key) && !isId(item.key));
```
Comment (nolleto) [inline app/assets/javascripts/models/beta/history.js]: IMHO, [reject](https://underscorejs.org/#reject) method sounds better in this scenario.
```js
export const filterChanges = changes =>
  reject(changes, item => isInitials(item.key) || isId(item.key));
```

File: app/assets/javascripts/models/beta/history.js (hunk 22 above)
Comment (edumoreira1506) [reply app/assets/javascripts/models/beta/history.js]: Nice. Thanks!
